### PR TITLE
Subscription improvements

### PR DESCRIPTION
**Pull request by DJSanti**
## What was changed
Significant revisions were made to all areas of the code, including:
- Simplifying the data structs and Activities
- Changing the loop in the Workflow Definition to see if the user is still subscribed
- Adding failure tests
- Changing variables and messaging to be closer to the Python example

## Why?
These apps and tutorials need to be more similar. I was diverging and needed to reign it in.

## Checklist
See above.

## How was this tested:
- Tested myself via web browser and Temporal Web UI
- Ran tests in test file

## Any docs updates needed?
The tutorial for Subscription in Go will need to be revised.

### CODE REVIEW

File: activities.go [hunk 1]
```diff
@@ -3,32 +3,14 @@ package subscribe_emails
...
 
-type Activities struct {
-}
-
 // email activities
-func (a *Activities) SendWelcomeEmail(ctx context.Context, emailInfo EmailInfo) (string, error) {
-	activity.GetLogger(ctx).Info("sending welcome email to customer", "EmailAddress", emailInfo.EmailAddress)
-	return "Sending welcome email completed for " + emailInfo.EmailAddress, nil
+func SendEmail(ctx context.Context, emailInfo EmailDetails) (string, error) {
+	activity.GetLogger(ctx).Info("Sending email to the following customer: %v", emailInfo.EmailAddress)
```
Comment (cretz) [inline activities.go:13]: The previous way the log was written is better. This log statement doesn't support `%` formatting.

**Review by cretz — COMMENTED** [review]
Added some first-glance comments. Looked at Go syntax and not really design first.

File: workflow.go [hunk 2]
```diff
@@ -11,28 +11,24 @@ import (
...
-	subscriptionPeriod := 0
-	// duration can be set up to a month.
-	duration := 10 * 24 * time.Hour
-
+func SubscriptionWorkflow(ctx workflow.Context, emailDetails EmailDetails) error {
+	subscriptionPeriodCount := emailDetails.SubscriptionCount
+	duration := 5 * time.Second
 	logger := workflow.GetLogger(ctx)
-	logger.Info("Subscription created", "EmailAddress", subscription.EmailInfo.EmailAddress)
+	logger.Info("Subscription created for: ", emailDetails.EmailAddress)
```
[the remark this answers is not in the document]
Comment (DJSanti) [reply workflow.go:18]: Will fix.

File: workflow.go [hunk 3]
```diff
@@ -46,66 +42,83 @@ func SubscriptionWorkflow(ctx workflow.Context, subscription Subscription) error
...
 	if err != nil {
 		return err
 	}
 
 	// start subscription period. execute until MaxBillingPeriods is reached
-	for ; subscriptionPeriod < subscription.MaxSubscriptionPeriods; subscriptionPeriod++ {
-		data := EmailInfo{
-			EmailAddress: subscription.EmailInfo.EmailAddress,
-			Mail:         "This is yet another email in the Subscription Workflow.",
+	for ; emailDetails.IsSubscribed ; {
```
[the remark this answers is not in the document]
Comment (DJSanti) [reply workflow.go:99]: This is false if the MaxCount is hit or the user cancels

[the remark this answers is not in the document]
Comment (DJSanti) [reply workflow.go:99]: Okay, I see where I messed it up. I'm going to have it compare the count to the maxcount instead

**Pull request merged**